docs/proposals/volume-idempotency.md

```diff
@@ -0,0 +1,229 @@
+    of plugins
+
+The volume `Attributes` structure is the appropriate place to store information the properties we
+discuss here:
+
+```go
+package volume
+
+type Attributes struct {
+	// other fields omitted
+
+	IdempotencyStrategy IdempotencyStrategy
+	Refreshable          bool
+}
+
+type IdempotencyStrategyType string
+
+const (
+    IdempotencyStrategyNone       IdempotencyStrategyType = "none"
+    IdempotencyStrategyMountpoint IdempotencyStrategyType = "mountPoint"
+    IdempotencyStrategyMarkerFile IdempotencyStrategyType = "markerFile"
+)
+```
+
+At runtime, these checks are represented by an interface:
+
+```go
+package volume
+
+type IdempotencyStrategy interface {
+    // ShouldRun returns (ok-to-run-setup, error)
+    ShouldRun() (bool, error)
+    Mark() error
+}
+```
+
+Refreshable plugins are represented by a new interface:
+
+```go
+package volume
+
+type RefreshableBuilder {
+    Builder
+
+    // Refresh returns (did-state-change, error)
+    Refresh(dir string) (bool, error)
+}
+
+```
+
+The workflow from the Kubelet's perspective for handling volume setup and refresh is:
+
+```go
+// go-ish pseudocode
+func mountExternalVolumes(pod) error {
+    podVolumes := make(kubecontainer.VolumeMap)
+    for i := range pod.Spec.Volumes {
+        volSpec := &pod.Spec.Volumes[i]
+        hasFSGroup := false
+        var fsGroup int64 = 0
+        if pod.Spec.SecurityContext != nil &&
+            pod.Spec.SecurityContext.FSGroup != nil {
+            hasFSGroup = true
+            fsGroup = *pod.Spec.SecurityContext.FSGroup
+        }
+
+        // Try to use a plugin for this volume.
+        plugin := volume.NewSpecFromVolume(volSpec)
+        builder, err := kl.newVolumeBuilderFromPlugins(plugin, pod)
+        if err != nil {
+            return err
+        }
+        if builder == nil {
+            return errUnsupportedVolumeType
+        }
+
+        var (
+            runSetup               = false
+            tryOwnershipManagement = false
+            idempotencyStrategy    = makeIdempotencyStrategy(builder.GetAttributes().IdempotencyStrategy)
+        )
+
+        runSetup, err := idempotencyStrategy.ShouldRun()
+        if err != nil {
+            return err
+        }
+        if runSetup {
+            err = builder.SetUp()
+            if err == nil {
+                idempotencyStrategy.Mark()
+                tryOwnershipManagement = true
+            } else {
+                return err
+            }            
+        } else if builder.GetAttributes().Refreshable {
+            refreshable, ok := builder.(RefreshablePlugin)
+            if !ok {
+                return errNotRefreshable
+            }
+
+            didRefresh, err := refreshable.Refresh()
+            if err != nil {
+                return err
+            }
+
+            if didRefresh {
+                tryOwnershipManagement = true
+            }
+        }
```

Comment:
what if `SetUp` returns a boolean in addition to the error. True if it is a new volume/false if it is an old volume. Plugins can decide what to return in that boolean. Host path always returns false; volumes with a mount dir use the existence of the dir to decide; NFS GlusterFS return always false; empty_dir and ilk return the value of `IsReady()`; block device volumes return a value based on whether they were formatted druing `SetUp()`